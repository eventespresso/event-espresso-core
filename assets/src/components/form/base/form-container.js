/**
 * Internal imports
 */
import './form-container.css';

/**
 * External imports
 */
import classNames from 'classnames';


/**
 * FormContainer is a wrapper that hides the form while it is loading,
 * then fades it in once loading is complete
 *
 * @function
 * @param {Object} form
 * @param {string} htmlClass
 * @param {boolean} loading
 * @return {Object} form
 */
export const FormContainer = ( { children, htmlClass, loading } ) => {
	const containerClass = classNames( {
		'ee-form-container-div': true,
		[ `${ htmlClass }` ]: htmlClass,
		'ee-form-fade-in': ! loading,
	} );
	return (
		<div className={ containerClass } >
			{ ! loading && children }
		</div>
	);
};
