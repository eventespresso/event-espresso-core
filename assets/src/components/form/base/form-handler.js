/**
 * External imports
 */
import { Form } from 'react-final-form';
import { castArray, isEqual } from 'lodash';
import { useEffect, useState } from '@wordpress/element';

/**
 * Internal imports
 */
import { FormErrorBoundary } from './form-error-boundary';
import FormHandlerForm from './form-handler-form';

const nullFunc = () => null;

/**
 * FormHandler
 * Higher-Order-Component that wraps the supplied form component
 * with a React-Final-Form Form component and connects it to the
 * supplied load, Submit, and Reset handlers, which are
 * passed as props to the final wrapped component.
 *
 * @param {Function} FormComponent
 * @return {Object} FormComponent with added form handling
 */
const FormHandler = ( {
	loadHandler = nullFunc,
	submitHandler = nullFunc,
	setMutatorCallbacks = nullFunc,
	submitButtonText = '',
	cancelButtonText = '',
	errorMessage = '',
	loadingNotice = '',
	decorators = [],
	mutators = {},
	...otherProps
} ) => {
	const [ formData, setFormData ] = useState( {} );
	useEffect(
		() => setFormData( loadHandler() ),
		[ loadHandler ]
	);
	return (
		<FormErrorBoundary errorMessage={ errorMessage } >
			<Form
				{ ...otherProps }
				onSubmit={ submitHandler || nullFunc }
				initialValues={ formData }
				decorators={ castArray( decorators ) }
				mutators={ mutators }
				render={ ( {
					form,
					values,
					initialValues,
					...formProps
				} ) => {
					setMutatorCallbacks( form.mutators );
					return (
						<FormHandlerForm
							{ ...formProps }
							form={ form }
							initialValues={ initialValues }
							currentValues={ values }
							notice={ loadingNotice }
							showSubmit={ submitHandler !== nullFunc }
							submitButtonText={ submitButtonText }
							cancelButtonText={ cancelButtonText }
							pristine={ isEqual( values, initialValues ) }
						/>
					);
				} }
			/>
		</FormErrorBoundary>
	);
};

export default FormHandler;
