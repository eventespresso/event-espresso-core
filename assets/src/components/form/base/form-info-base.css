
.ee-form-info-row {
    margin: var(--ee-margin-small) 0;
}

.ee-form-info-row.is-dismissable {
    max-height: 500px;
    min-height: 0 !important;
    overflow: hidden;
    transition: max-height 0.25s ease-in-out;
}

.ee-form-info-row.is-dismissable.dismissed {
    max-height: 0 !important;
}

.ee-form-info {
    font-style: italic;
    height: auto;
    line-height: 1.25rem;
    min-height: var(--ee-icon-button-size);
    position: relative;
}

.ee-form-info-text {
    height: 100%;
}

.ee-form-info-row.is-dismissable .ee-form-info {
    padding: 0 var(--ee-padding-small) 0 0;
}

.ee-form-info .ee-form-info-type {
    background: var(--ee-background-color);
    color: var(--ee-color-grey-10);
    left: calc(var(--ee-padding-extreme) * -1);
    position: absolute;
    top: 0;
}

.ee-form-info.ee-form-info-type-warning .ee-form-info-type {
    color: var(--ee-color-hot-pink);
    font-size: var(--ee-font-size-default);
}

.ee-form-info .ee-form-info-text .dashicon {
    margin: 0 var(--ee-margin-micro);
    vertical-align: text-bottom;
}

.ee-form-info .ee-form-info-dismiss {
    background: var(--ee-background-color);
    color: var(--ee-color-grey-10);
    padding: var(--ee-padding-micro);
    position: absolute;
    right: 0;
    top: 0;
}

.ee-form-info .ee-form-info-dismiss:hover,
.ee-form-info .ee-form-info-dismiss:focus,
.ee-form-info .ee-form-info-dismiss:active {
    box-shadow: none !important;
    color: var(--ee-default-text-color-low-contrast) !important;
    outline: none !important;
}

.dismissed .ee-form-info .ee-form-info-dismiss:hover {
    animation: var(--ee-animation-spin-fast);
}


@media screen and (max-width: 600px) {
    .ee-form-info .ee-form-info-type {
        left: var(--ee-padding-tiny);
    }

    .ee-form-info-text {
        display: inline-block;
        padding-left: var(--ee-padding-extreme);
    }

}
