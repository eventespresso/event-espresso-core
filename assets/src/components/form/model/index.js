export {
	ModelSelect,
	default as ModelEnhancedSelect,
} from './base/model-select';
export * from './select';
export * from './base/build-options';
export { default as createModelSelect } from './base/create-model-select';
