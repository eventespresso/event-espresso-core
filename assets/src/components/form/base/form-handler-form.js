/**
 * External imports
 */
import { isEmpty } from 'lodash';
import { useCallback, useMemo } from '@wordpress/element';

/**
 * Internal imports
 */
import { FormContainer } from './form-container';
import { FormDataDebugDump } from './form-data-debug-dump';
import FormPlaceholder from './form-placeholder';
import FormSubmitButton from './form-submit-button';
import FormCancelButton from './form-cancel-button';

const nullFunc = () => null;

const FormHandlerForm = ( {
	FormComponent,
	form,
	initialValues,
	currentValues,
	handleSubmit,
	submitting,
	pristine,
	invalid,
	loading,
	loadingNotice,
	showSubmit,
	submitButtonText,
	showCancel,
	cancelButtonText,
	formChanges = false,
	resetHandler = nullFunc,
	...formProps
} ) => {
	pristine = pristine && ! formChanges;
	currentValues = isEmpty( currentValues ) ? initialValues : currentValues;
	const formReset = useCallback(
		( event ) => {
			if ( resetHandler !== nullFunc ) {
				resetHandler( event );
			}
			form.reset( event );
		},
		[ resetHandler, form.reset ]
	);
	const submitButton = useMemo(
		() => showSubmit ?
			<FormSubmitButton
				buttonText={ submitButtonText }
				submitting={ submitting }
				disabled={ pristine || invalid }
			/> :
			null,
		[ showSubmit, submitButtonText, submitting, pristine, invalid ]
	);
	const cancelButton = useMemo(
		() => showCancel ?
			<FormCancelButton
				buttonText={ cancelButtonText }
				onClick={ formReset }
				pristine={ pristine }
				submitting={ submitting }
			/> :
			null,
		[ showCancel, cancelButtonText, formReset, submitting, pristine ]
	);
	return (
		<form onSubmit={ handleSubmit }>
			<FormPlaceholder
				loading={ loading }
				notice={ loadingNotice }
			/>
			<FormContainer loading={ loading }>
				<FormComponent
					updateField={ form.change }
					touchField={ form.blur }
					initialValues={ initialValues }
					currentValues={ currentValues }
					submitButton={ submitButton }
					cancelButton={ cancelButton }
					formReset={ formReset }
					pristine={ pristine }
					{ ...formProps }
				/>
			</FormContainer>
			<FormDataDebugDump values={ currentValues } />
		</form>
	);
};

export default FormHandlerForm;
