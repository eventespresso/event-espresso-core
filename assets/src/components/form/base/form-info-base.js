/**
 * External imports
 */
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import { __ } from '@eventespresso/i18n';
import { Dashicon, IconButton } from '@wordpress/components';

/**
 * Internal dependencies
 */
import './form-info-base.css';

/**
 * FormInfoBase
 * displays instructions or other important information
 * that users may require to properly complete a form.
 * !!! IMPORTANT !!!
 * if the form info needs to be dismissable, then the
 * component wrapping this (ie: FormRow) should do the following:
 *  - have an html CSS class of "ee-form-info-row"
 *  - manage the "dismiss" state for the component
 *    which can be toggled via the onDismiss callback
 *  - add an html css class of "dismissed" to hide the form info
 *
 * @function
 * @param {Object} props
 * @member {string|Object} formInfo
 * @member {string} dashicon
 * @member {string} htmlClass
 * @member {Function} onDismiss
 * @member {Array} formInfoVars
 */
const FormInfoBase = ( {
	formInfo,
	dashicon,
	htmlClass,
	onDismiss,
	formInfoVars,
} ) => {
	htmlClass = htmlClass ?
		`${ htmlClass } ee-form-info` :
		'ee-form-info';
	htmlClass = dashicon ?
		`${ htmlClass } ee-form-info-type-${ dashicon } has-info-type` :
		htmlClass;
	const typeIcon = dashicon ? (
		<Dashicon
			className="ee-form-info-type"
			icon={ dashicon }
			size={ 36 }
		/>
	) : null;
	const dismiss = onDismiss ? (
		<IconButton
			className="ee-form-info-dismiss clickable"
			icon="dismiss"
			onClick={ onDismiss }
		/>
	) : null;
	// parseHtmlPlaceholders parse-html-placeholders
	if ( ! isEmpty( formInfoVars ) ) {
		const formInfoText = [];
		const chunks = formInfo.split( '%%var%%' );
		if ( chunks.length !== formInfoVars.length ) {
			throw new RangeError(
				'The number of notice placeholders does not match' +
				' the number of form info variables supplied.'
			);
		}
		for ( let x = 0; x < chunks.length; x++ ) {
			if ( chunks[ x ] ) {
				formInfoText.push( chunks[ x ] );
			}
			if ( formInfoVars[ x ] ) {
				formInfoText.push( formInfoVars[ x ] );
			}
		}
		formInfo = formInfoText;
	}
	return formInfo ? (
		<div
			aria-label={ __( 'important information', 'event_espresso' ) }
			className={ htmlClass }
		>
			{ typeIcon }
			<div className="ee-form-info-text">{ formInfo }</div>
			{ dismiss }
		</div>
	) : null;
};

FormInfoBase.propTypes = {
	formInfo: PropTypes.oneOfType( [
		PropTypes.string,
		PropTypes.object,
	] ).isRequired,
	dashicon: PropTypes.string,
	htmlClass: PropTypes.string,
	onDismiss: PropTypes.func,
	formInfoVars: PropTypes.array,
};

FormInfoBase.defaultProps = {
	dashicon: '',
	htmlClass: '',
	onDismiss: null,
	formInfoVars: [],
};

export default FormInfoBase;
