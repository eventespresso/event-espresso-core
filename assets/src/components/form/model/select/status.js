/**
 * Internal imports
 */
import createModelSelect from '../base/create-model-select';
import { statusModel as model } from '../../../../data/model';
import { withBaseControl } from '../../../../higher-order-components';

/**
 * External imports
 */
import { __ } from '@eventespresso/i18n';
import PropTypes from 'prop-types';
import { statusModel } from '@eventespresso/model';

const optionsEntityMap = {
	default: {
		value: 'STS_ID',
		label: (statusEntity) => statusModel.prettyStatus(statusEntity.STS_ID),
	},
};

/**
 * @type {ModelSelectComponent}
 */
const StatusSelect = createModelSelect(
	model.MODEL_NAME,
	{
		selectConfiguration: {
			loadingMessage: () => __('Retrieving Statuses.', 'event_espresso'),
			noOptionsMessage: () => __('No Statuses.', 'event_espresso'),
			placeholder: __('Select Statusâ€¦', 'event_espresso'),
		},
		...model.defaultQueryData,
		getQueryString: model.getQueryString,
		label: __('Select Status', 'event_espresso'),
		optionsEntityMap,
	},
	{
		...model.queryDataTypes,
		selected: PropTypes.oneOf(model.ALL_STATUS_IDS),
		optionsEntityMap: PropTypes.object,
	}
);

export default StatusSelect;

/**
 * Enhanced Status Select for the WordPress editor
 */
export const EditorStatusSelect = withBaseControl('select-status')(
	StatusSelect
);
