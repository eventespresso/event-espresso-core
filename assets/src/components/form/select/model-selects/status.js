/**
 * Internal imports
 */
import ModelSelect from '../model-select';
import { statusModel as model } from '../../../../data/model';
import { withBaseControl } from '../../../../higher-order-components';

/**
 * External imports
 */
import { __ } from '@eventespresso/i18n';
import { Component } from '@wordpress/element';
import { PropTypes } from 'prop-types';

const optionsEntityMap = {
	default: {
		value: 'STS_ID',
		label: 'STS_code',
	},
};

/**
 * Select Component for the Status Model.
 */
export default class StatusSelect extends Component {
	state = {
		modelName: model.MODEL_NAME,
		queryData: {},
	};

	static defaultProps = {
		selectConfiguration: {
			loadingMessage: () => __( 'Retrieving Statuses.',
				'event_espresso',
			),
			noOptionsMessage: () => __( 'No Statuses.', 'event_espresso' ),
			placeholder: __( 'Select Status...', 'event_espresso' ),
		},
		...model.defaultQueryData,
		getQueryString: model.getQueryString,
		label: __( 'Select Status', 'event_espresso' ),
		optionsEntityMap,
	};

	static propTypes = {
		...model.queryDataTypes,
		statusType: PropTypes.oneOf( [
			model.STATUS_TYPE_EMAIL,
			model.STATUS_TYPE_EVENT,
			model.STATUS_TYPE_MESSAGE,
			model.STATUS_TYPE_PAYMENT,
			model.STATUS_TYPE_REGISTRATION,
			model.STATUS_TYPE_TRANSACTION,
		] ).isRequired,
		selectedStatusId: PropTypes.oneOf(
			model.ALL_STATUS_IDS
		),
		onStatusSelect: PropTypes.func,
		label: PropTypes.string,
		optionsEntityMap: PropTypes.object,
	};

	addStatusTypeToQueryData( statusType ) {
		this.setState( {
			queryData: {
				...this.state.queryData,
				statusType,
			},
		} );
	}

	componentDidMount() {
		this.setState( {
			queryData: {
				...this.props.queryData,
				statusType: this.props.statusType,
			},
		} );
	}

	componentDidUpdate( prevProps ) {
		if ( prevProps.statusType !== this.props.statusType ) {
			this.addStatusTypeToQueryData( this.props.statusType );
		}
	}

	render() {
		const { selectedStatusId, onStatusSelect } = this.props;
		const selectOpts = {
			selectConfiguration: {
				defaultValue: selectedStatusId,
				onChange: onStatusSelect,
				...this.props.selectConfiguration,
			},
		};
		const props = {
			...this.props,
			...selectOpts,
			...this.state,
		};
		return <ModelSelect { ...props } />;
	}
}

/**
 * Enhanced Status Select for the WordPress editor
 */
export const EditorStatusSelect = withBaseControl(
	'select-status'
)( StatusSelect );
