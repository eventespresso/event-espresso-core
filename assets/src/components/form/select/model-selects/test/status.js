import StatusSelect, { EditorStatusSelect } from '../status';
import { shallow, render } from 'enzyme';
import { QUERY_ORDER_DESC } from 'assets/src/data/model/base';
import * as statusModel from 'assets/src/data/model/status';

jest.mock( '../../model-select', () => () => <span>StatusSelect</span> );

describe( 'StatusSelect()', () => {
	const expectedDefaultQueryData = {
		limit: 25,
		orderBy: 'statusCode',
		order: QUERY_ORDER_DESC,
		statusType: statusModel.STATUS_TYPE_REGISTRATION,
	};
	it( 'matches snapshot with default props plus provided statusType', () => {
		const wrapper = shallow(
			<StatusSelect
				statusType={ statusModel.STATUS_TYPE_REGISTRATION }
			/>
		);
		expect( wrapper ).toMatchSnapshot();
	} );
	it( 'matches expected default props plus provided statusType', () => {
		const wrapper = shallow(
			<StatusSelect
				statusType={ statusModel.STATUS_TYPE_REGISTRATION }
			/>
		);
		expect( wrapper.prop( 'queryData' ) )
			.toEqual( expectedDefaultQueryData );
		expect( wrapper.prop( 'selectConfiguration' ).placeholder ).toEqual(
			'Select Status...',
		);
		expect( wrapper.prop( 'modelName' ) ).toEqual( 'status' );
		expect( wrapper.prop( 'label' ) ).toEqual( 'Select Status' );
	} );
} );

describe( 'EditorStatusSelect()', () => {
	it( 'is expected to fail because of missing required prop', () => {
		render( <EditorStatusSelect /> );
		expect( console ).toHaveErrored();
	} );
	it( 'matches snapshot with default props', () => {
		const wrapper = render(
			<EditorStatusSelect
				statusType={ statusModel.STATUS_TYPE_REGISTRATION }
			/>
		);
		expect( wrapper ).toMatchSnapshot();
	} );
} );
