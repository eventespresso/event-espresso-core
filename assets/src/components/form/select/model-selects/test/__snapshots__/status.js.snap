// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorStatusSelect() matches snapshot with default props 1`] = `
<div
  class="components-base-control"
>
  <div
    class="components-base-control__field"
  >
    <label
      class="components-base-control__label"
      for="inspector-select-status-control-1"
    >
      Select Status
    </label>
    <span>
      StatusSelect
    </span>
  </div>
</div>
`;

exports[`StatusSelect() matches snapshot with default props plus provided statusType 1`] = `
<Component
  getQueryString={[Function]}
  label="Select Status"
  modelName="status"
  optionsEntityMap={
    Object {
      "default": Object {
        "label": "STS_code",
        "value": "STS_ID",
      },
    }
  }
  queryData={
    Object {
      "limit": 25,
      "order": "DESC",
      "orderBy": "statusCode",
      "statusType": "registration",
    }
  }
  selectConfiguration={
    Object {
      "defaultValue": undefined,
      "loadingMessage": [Function],
      "noOptionsMessage": [Function],
      "onChange": undefined,
      "placeholder": "Select Status...",
    }
  }
  statusType="registration"
/>
`;
